### PR TITLE
Add support for MHLO scatter op lowering to LinalgExt when the indices have an implicit trailing dim

### PR DESCRIPTION
**Pull request by NatashaKnk**
Fix for [issue 9669](https://github.com/iree-org/iree/issues/9669)

### CODE REVIEW

File: compiler/src/iree/compiler/InputConversion/MHLO/ConvertMHLOToLinalgExt.cpp [hunk 1]
```diff
@@ -205,13 +205,18 @@ struct ScatterOpConversion : public OpConversionPattern<mhlo::ScatterOp> {
...
-
-    if (indicesRank < 2) return false;
-    if (dimNumbers.getIndexVectorDim() != indicesRank - 1) return false;
-
+    auto indexVectorDim = dimNumbers.getIndexVectorDim();
     auto indexDepth = indicesType.getShape().back();
     auto scatterDimsToOperandDims = dimNumbers.getScatterDimsToOperandDims();
-    if (scatterDimsToOperandDims.size() != indexDepth) return false;
+
+    // The indices have an implicit trailing dim
```
Comment (jpienaar) [inline compiler/src/iree/compiler/InputConversion/MHLO/ConvertMHLOToLinalgExt.cpp:212]: Could you turn this into a full sentence?

File: compiler/src/iree/compiler/InputConversion/MHLO/ConvertMHLOToLinalgExt.cpp [hunk 2]
```diff
@@ -228,6 +233,37 @@ struct ScatterOpConversion : public OpConversionPattern<mhlo::ScatterOp> {
...
+      mhlo::ScatterOp op, ConversionPatternRewriter &rewriter,
+      OpAdaptor adaptor) {
+    auto scatterIndices = adaptor.scatter_indices();
+    auto scatterDims = adaptor.scatter_dimension_numbers();
+    auto indicesType = scatterIndices.getType().cast<ShapedType>();
+    auto indicesShape = indicesType.getShape();
+    auto indicesRank = indicesType.getRank();
+    auto indexVectorDim = scatterDims.getIndexVectorDim();
+
+    if (indexVectorDim == indicesRank) {
```
Comment (jpienaar) [inline compiler/src/iree/compiler/InputConversion/MHLO/ConvertMHLOToLinalgExt.cpp:246]: Is only 1 trailing dim implicit dim allowed?

**Review by jpienaar — COMMENTED** [review]
Nice, could you add a test too?

File: compiler/src/iree/compiler/InputConversion/MHLO/ConvertMHLOToLinalgExt.cpp (hunk 2 above)
Comment (NatashaKnk) [reply compiler/src/iree/compiler/InputConversion/MHLO/ConvertMHLOToLinalgExt.cpp:246]: As far as I understand, yes. (Plus, if it's implicit - how would we know to differentiate between one or more? :)

**Comment by NatashaKnk** [comment]
What's the standard for adding tests for mhlo -> LinalgExt? I couldn't find a dedicated tests folder @rsuderman

**Comment by hanhanW** [comment]
> What's the standard for adding tests for mhlo -> LinalgExt? I couldn't find a dedicated tests folder @rsuderman

mhlo -> LinalgExt lit tests are in https://github.com/iree-org/iree/blob/main/compiler/src/iree/compiler/InputConversion/MHLO/test/convert_mhlo_to_linalg_ext.mlir

**Review by MaheshRavishankar — APPROVED** [review]

File: compiler/src/iree/compiler/InputConversion/MHLO/ConvertMHLOToLinalgExt.cpp (hunk 2 above)
Comment (rsuderman) [reply compiler/src/iree/compiler/InputConversion/MHLO/ConvertMHLOToLinalgExt.cpp:247]: 👍

**Review by rsuderman — APPROVED** [review]

**Review by jpienaar — APPROVED** [review]
Nice, thanks!

**Comment by GMNGeoffrey** [comment]
Please remember to use "squash and merge" when the commit history on your branch is messy or not interesting. Things like "fix comment" and "Merge branch 'iree-org:main' into main" are now in the permanent git history :-/

**Pull request merged**